### PR TITLE
fix(iota-genesis-builder): Timelock allocations execution

### PR DESCRIPTION
**Pull request by miker83z**
# Description of change

Fixes some issues of #741 

[edit]

Main issue fixed with this PR:
- There was an issue with the execution: timelock staking requires `TimeLocks` objects and the `IotaSystemState` shared object as inputs to the PTB. SInce the main genesis creation PTB does not take inputs and since the `IotaSystemState` shared object cannot be referenced in the same PTB (without changing the Move implementation of `genesis::create`) a second PTB was needed to be executed just for timelocks allocation. This is implemented with [e450ead](https://github.com/iotaledger/iota/pull/918/commits/e450ead69f249b2eabcac9577e79b454becd0280).

Minor issues:
- Another issue was due to the use of some timelocks that are already expired. Fixed with #925 
- When a stake is initialized for each validator [`init_voting_power_info` is called](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-framework/packages/iota-system/sources/voting_power.move#L68); if only timelocks are allocated (i.e., no gas coins), during the `genesis::create` then [this line will divide by zero](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-framework/packages/iota-system/sources/voting_power.move#L80) because `total_stake` is zero. TEMPORARILY fixed with [bba27d7](https://github.com/iotaledger/iota/pull/918/commits/bba27d7f2e57785c92d8c1c47687715f40a7bb80). Issue tracked by #927  
- After the execution was running correctly, another issue appeared because the genesis stake `timelock_allocation` was taken out from its memory location for the PTB execution and then dropped; however, it was needed to execute the last `validate()` [at the end of the build](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-genesis-builder/src/lib.rs#L261); fixed with [0112717](https://github.com/iotaledger/iota/pull/918/commits/011271768e989408211149e5bd700158a19068a5)
- Finally `validate_output()` is based on the assumption that the genesis is built with version 1 of the `IotaSystemState`; fixed with [fcb6b37](https://github.com/iotaledger/iota/pull/918/commits/fcb6b37a22e3be5cc344b19ed6e3d06e8db03910).

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

`cargo run --release --example build_stardust_genesis`

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by kodemartin — CHANGES_REQUESTED** [review]
Hey @miker83z, I reckon that this patch does a great job in uncovering a few issues that we have to tackle eventually.

But at the same time I feel it is important to distinguish between what is a fix and what is actually a workaround that can accommodate a successful local run on the basis of using a particular input. For example

> Another issue was due to the use of some timelocks that are already expired so I had to temporarely relax a related check in request_add_stake with [fbcdaf4](https://github.com/iotaledger/iota/pull/741/commits/fbcdaf446ef57a5184d8f061f07b76e907f377fe) while waiting for the fix of
https://github.com/iotaledger/iota/issues/904

This is a workaround that deals with timelocks that ought to have been coins. Otherwise, if we expect such expirations we might as well adjust the expiration check on the Move side.

> When a stake is initialized for each validator [init_voting_power_info is called](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-framework/packages/iota-system/sources/voting_power.move#L68); if only timelocks are allocated (i.e., no gas coins), during the genesis::create then [this line will divide by zero](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-framework/packages/iota-system/sources/voting_power.move#L80) because total_stake is zero. Fixed with [2715afb](https://github.com/iotaledger/iota/pull/741/commits/2715afb4a4e81a0ce93a3467f69d3bc4fb9e2da1). However this is temporary because allocates a lot of coins (30 million timelocked and 30 million gas coins). We need to come up with another strategy.

This is also a workaround, as the allocation implemented is not optimal as you point out.

I believe that it would be best to not include these workarounds in the final patch.  This requires actual removal and pin-pointing the code/logic to remove, and injects logical suprises while reading the current implementation. 

Instead we can create an issue for the `division by zero` problem and deal with that independently.

Furthermore,

> After the execution was running correctly, another issue appeared because the genesis stake timelock_allocation was taken out from its memory location for the PTB execution and then dropped; however, it was needed to execute the last validate() [at the end of the build](https://github.com/iotaledger/iota/blob/bf27f2f34073fedf92cf21ee8a515ec5fef3d56b/crates/iota-genesis-builder/src/lib.rs#L261); fixed with [4f2f2d8](https://github.com/iotaledger/iota/pull/741/commits/4f2f2d856dd8d9a794116cef37ff6f16444e86c6)

this is only an issue because you introduced the `validate_token_distribution_schedule` as part of the `validate` side effects. Upon revisiting, I have to wonder why is this necessary, since we have already this check in 

https://github.com/iotaledger/iota/blob/bd26be4a5dec5144213208b3d12df7ab5ae7efa8/crates/iota-genesis-builder/src/lib.rs#L239

and we could move it to L255 ensuring that the token-distribution schedule is valid. Avoiding this check in `Builder::validate` means that we won't have to deal with this side effect in other calls of the method as in the genesis-ceremony.

To recap I suggest the following:

* Revert local workarounds employed to pass the manual tests locally.
* Create issues for problems found in the testing phase. 
* Avoid unnecessary side-effect in existing methods.

**Review by kodemartin — APPROVED** [review]
lgtm :sunflower:

**Review by lzpap — APPROVED** [review]
